Add hcloud server module

LKaemmerling opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
As follow up from https://github.com/ansible/ansible/pull/52553 we would like to contribute the first module for our Hetzner Cloud: `hcloud_server`
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
hcloud_server

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

LKaemmerling committed:
Add hcloud_server module

LKaemmerling committed:
Add missing doc fragment

resmo commented on lib/ansible/module_utils/hcloud.py:
```diff
@@ -0,0 +1,60 @@
+# Copyright: (c) 2019, Hetzner Cloud GmbH <info@hetzner-cloud.de>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
code in module_utils should not be GNU licensed by convention.
~~~
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
~~~

resmo commented on lib/ansible/module_utils/hcloud.py:
```diff
@@ -0,0 +1,60 @@
+        self.represent = represent
+        self.result = {"changed": False, self.represent: None}
+        if not HCLOUD_AVAILABLE:
+            module.fail_json(msg="hcloud-python must be available to use this module")
```
there has been some improvement how to fail in this case. see https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/cloudstack.py#L60 as an example

resmo commented on lib/ansible/module_utils/hcloud.py:
```diff
@@ -0,0 +1,60 @@
+    def base_module_arguments():
+        return {
+            "api_token": {
+                "type": "str",
```
please add a `no_log: True` do not show secrets and passwords in the logs.

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,332 @@
+            - Resize the disk size, when resizing a server. If you want to downgrade the server later, this value should
+            - be False.
+        type: bool
+        default: False
```
minor, convention is to use `yes` or  `no`, and not `True` or `False`, for values in case type: bool

dagwieers replied:
I prefer yes/no too, but a recent docs meeting allowed for either yes/no or true/false in documentation. In this case the documentation is being processed and will end up being yes/no anyway, so it doesn't matter that much.

However for EXAMPLES it does matter as they appear in the documentation as-is, and yes/no is recommended for consistency. So I would advise to use yes/no as much as possible (even if it is not user-facing).

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,332 @@
+        default: False
+    upgrade_disk:
+        description:
+            - Resize the disk size, when resizing a server. If you want to downgrade the server later, this value should
```
I would put each phrase on a line item

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,332 @@
+    ssh_keys:
+        description:
+            - List of SSH Keys Names
+        type: list
```
ansible has some magic and allows this for type list

~~~yaml
ssh_keys: mykey
~~~

which is identical to 

~~~yaml
ssh_keys: [ mykey ]
~~~

therefore I would put an `aliases: [ ssh_key ]`here and in the arg_spec to make it more conventient: 

~~~yaml
ssh_key: mykey
~~~

dagwieers replied:
I disagree, if we end up with people using `ssh_key: my_key` in their playbooks, people may assume only a single key can be added. Keeping only the plural `ssh_keys` form will make clear more than one SSH key is possible, even if only one is provided.

So adding the singular form for convenience introduces potential confusion and inconsistencies between playbooks and users. It's always better to keep the number of aliases to a minimum and only add them for backward compatibility or to add consistency (e.g. using `name` or `path` for abstraction).

Besides it would mean we do this for every plural parameter name (usually lists) which is definitely not true.

dagwieers replied:
cc acozine gundalow Should we clarify this in the module development docs ?

resmo replied:
it is a user interface, if users don't read docs it can't be our fault. anyways. not a blocker (that is why I used "would")

dagwieers replied:
@resmo But if adding something for convenience can cause confusion, it may not be in our best interest. Good user interface design should not require more documentation, it should make documentation less important. The ultimate goal would be to not need any documentation because things speak for themselves, and I think that is what `ssh_keys` does in this case.

samccann replied:
This was discussed in the ansible-docs meeting and we were in general agreement with @dagweirs here, esp clarifying when to use an alias in modules. We'll get that added to developer documentation in a separate PR.

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,332 @@
+    image: ubuntu-18.04
+    location: fsn1
+    ssh-keys:
+        - my-ssh-key
```
please add some more examples for e.g. for state=absent

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,332 @@
+    server_type:
+        description:
+            - The Server Type
+            - Required if server does not exists
```
not quite true, it would be required if state is not absent right?

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,332 @@
+    location:
+        description:
+            - Location of Server
+            - Required if no datacenter is given and server does not exists
```
to show datacenter is anoter param use I(datacenter)

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,332 @@
+version_added: "2.8"
+
+description:
+    - "Create and manage cloud servers on the Hetzner Cloud."
```
I would put an note about updating here

LKaemmerling committed:
Apply review results from resmo

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,376 @@
+    name:
+        description:
+            - The Server Name
+            - Only required if no server id is given or a server does not exists
```
typo: exists -> exist

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,376 @@
+    server_type:
+        description:
+            - The Server Type
+            - Required if server does not exists
```
typo: exists -> exist

LKaemmerling committed:
Apply review results from resmo
Introduce force_upgrade flag

decentral1se commented:
Closes https://github.com/ansible/ansible/issues/52813#issuecomment-468146009.

gandhiano reviewed (approved):
Properly designed and addressing the requirements of #52813 for provisioning and managing hetzner cloud hosts.

decentral1se commented:
Opening https://github.com/ansible/molecule/issues/1791 to cover on Molecule side :+1:

LKaemmerling committed:
Fix location and datacenter

LKaemmerling committed:
Improve Error Handling when Server does not exists.

resmo commented on test/integration/targets/hcloud_server/tasks/main.yml:
```diff
@@ -0,0 +1,242 @@
+      - main_server.hcloud_server.server_type == "cx11"
+      - main_server.hcloud_server.status == "running"
+      - main_server.root_password != ""
+
```
would you mind adding idempotency testst? basically same tasks check but with `main_server is not changed`?

resmo commented:
I see some improvements in handling the update scenario. There are several I would appreciate for consideration: 

Currently updating a running server would require to use `force_upgrade: yes` even if the target state is `state: stopped`. This is not necessary, the server could be stopped beforehand and updated without need of `force_upgrade`.

Updating a running server with a target state of `state: present` and `force_upgrade: yes` result in `state: stopped`. As a user I would expect that the server state would be kept (or restored after upgrade)

A similar logic is implemented in modules like vultr_server, or cloudstack_instance.

resmo commented:
needs_info

LKaemmerling committed:
Apply recent master changes
Allow resizing of server without force_upgrade when server is in state stopped
Start server after resize if state is not stopped

LKaemmerling commented:
@resmo I have changed the behavior when the server is in the state stopped and have applied your suggestions. I have rebased the branch, so the changes from https://github.com/ansible/ansible/pull/53044 are here done too.

LKaemmerling commented:
ready_for_review

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -306,6 +315,9 @@ def _update_server(self):
                     upgrade_disk=self.module.params.get("upgrade_disk"),
                 ).wait_until_finished(timeout)
+                if state != "stopped":
+                    self.start_server()
```
in case of `state: present` with a powered-off server it would be started here.
I would remember the server state before force-stop it and also extend the condition with it:

~~~diff
-                if state != "stopped":
+                if (state == "present" and previous_state == Server.STATUS_RUNNING) or state == "started":
                     self.start_server()
~~~

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -287,9 +295,10 @@ def _update_server(self):
             if self.hcloud_server.status == Server.STATUS_RUNNING:
                 if not self.module.check_mode:
-                    if self.module.params.get("force_upgrade"):
+                    if self.module.params.get("force_upgrade") or state == "stopped":
```
Remember the state before stop so it can be latter used.
~~~diff
                      if self.module.params.get("force_upgrade") or state == "stopped":
+                         previous_state = self.hcloud_server.status
~~~

resmo reviewed (commented):
needs_revision

LKaemmerling committed:
Start server after resize if it was present and running before

LKaemmerling commented:
ready_for_review

resmo commented:
LGTM

dagwieers commented on lib/ansible/module_utils/hcloud.py:
```diff
@@ -0,0 +1,61 @@
+# Copyright: (c) 2019, Hetzner Cloud GmbH <info@hetzner-cloud.de>
```
We tend to add this snippet at the top of every file (usually below the shebang if there is one).
```suggestion
# -*- coding: utf-8 -*-

# Copyright: (c) 2019, Hetzner Cloud GmbH <info@hetzner-cloud.de>
```

dagwieers commented on lib/ansible/module_utils/hcloud.py:
```diff
@@ -0,0 +1,61 @@
+try:
+    import hcloud
+
+    HCLOUD_AVAILABLE = True
```
The standard naming here is `HAS_HCLOUD`, not a real showstopper.

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+version_added: "2.8"
+
+description:
+    - "Create, update and manage cloud servers on the Hetzner Cloud."
```
No need for quotes here. We tend to only quote strings when it is required by YAML.
We do this so users have a better understanding when quoting is actually required.
This applies mostly to examples and documentation (not e.g. integration tests).

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+options:
+    id:
+        description:
+            - The Server ID
```
We tend to have full descriptions, so that the meaning is more clear to the end-user.
The first line is a description of the parameter. Subsequent lines clarify the value, its use or dependencies.
All descriptions (except the short_description) should have a trailing dot, so please add those everywhere.
```suggestion
            - The ID of the Hetzner server to manage.
```

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+    id:
+        description:
+            - The Server ID
+            - Only required if no server name is given
```
Please do this everywhere.
```suggestion
            - Only required if no server I(name) is given.
```

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+    name:
+        description:
+            - The Server Name
+            - Only required if no server id is given or a server does not exists
```
```suggestion
            - Only required if no server I(id) is given or a server does not exists.
```

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+        description:
+            - State of the server.
+        default: present
+        choices: [ present, absent, restarted, started, stopped ]
```
We tend to sort items in lists when the order is of no importance.
This helps people to compare lists, or find an entry more quickly.
The documentation sorts them as well during generation, so you cannot influence the order seen by the user.
```suggestion
        choices: [ absent, present, restarted, started, stopped ]
```

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+"""
+
+EXAMPLES = """
+# Create a basic server
```
There is not value here in having these comments if it contains the exact same information as in `name:`.
Please remove them from all the entries.
```suggestion
```

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+  hcloud_server:
+    name: my-server
+    server_type: cx21
+    keep_disk: true
```
We recommend to use yes/no values for booleans. Since these examples are shown as part of the module documentation and the documentation uses yes/no we prefer to keep it consistent within the documentation.

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+            "server_type": to_native(self.hcloud_server.server_type.name),
+            "datacenter": to_native(self.hcloud_server.datacenter.name),
+            "location": to_native(self.hcloud_server.datacenter.location.name),
+            "rescue_enabled": to_native(self.hcloud_server.rescue_enabled),
```
This would turn everything into a string, not sure if that's what you want for e.g. booleans).
Same for the other values that are not strings. Probably not what you want.

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+    state = module.params.get("state")
+    if state == "absent":
+        hcloud.delete_server()
+    else:
```
Why this additional indentation and not just a long if-elif-elif-elif-else chain ?

dagwieers commented on lib/ansible/plugins/doc_fragments/hcloud.py:
```diff
@@ -0,0 +1,25 @@
+options:
+    api_token:
+        description:
+            - This is the API Token for the Hetzner Cloud
```
Please add a trailing dot.

dagwieers commented on lib/ansible/plugins/doc_fragments/hcloud.py:
```diff
@@ -0,0 +1,25 @@
+        default: https://api.hetzner.cloud/v1
+        type: str
+requirements:
+  - "python >= 2.6"
```
No need for quotes here.

dagwieers commented on lib/ansible/plugins/doc_fragments/hcloud.py:
```diff
@@ -0,0 +1,25 @@
+        type: str
+requirements:
+  - "python >= 2.6"
+  - "hcloud-python >= 1.0.0"
```
No need for quotes here.

dagwieers commented on lib/ansible/plugins/doc_fragments/hcloud.py:
```diff
@@ -0,0 +1,25 @@
+  - "python >= 2.6"
+  - "hcloud-python >= 1.0.0"
+notes:
+  - Also see the API documentation on https://docs.hetzner.cloud/.
```
Please use a `seealso:` section (see https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-fields).
```suggestion
seealso:
- name: Documentation for Hetzner Cloud API
  description: Complete reference for the Hetzner Cloud API.
  link: https://docs.hetzner.cloud/
```

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+        description:
+            - List of SSH Keys Names
+        type: list
+        aliases: [ ssh_key ]
```
I would not add this alias. We tend to use aliases for backward compatibility or to add consistency between modules. But there's no value in adding this alias here. Rather stick to one parameter name.

LKaemmerling replied:
This was a suggestion from @resmo.

I think it is more intuitive for some people to write
```yaml
ssh_key: my-ssh-key
```
instead of 
```yaml
ssh_keys: my-ssh-key
```
or 
```yaml
    ssh_keys:
      - my-ssh-key
```

My suggestion would be to leave it, because some other cloud modules have the same behavior (vultr as sample)

dagwieers replied:
All our parameters that accept lists, accept single items too. But we never offer both the plural and singular version for parameter name. A one-item list, is still a list :-)

So I wouldn't bother. IMO offering both is more confusing as people may assume that:
```yaml
    ssh_key: my-ssh-key
```
means there is only one SSH key to add. That is why I would only have the plural form and no aliases. It will end up to be more consistent between users/playbooks and it's clear the parameter accepts more than one value.

dagwieers replied:
I clarified my position above, and this will spark a discussion. Don't worry, the best way to settle these cases once and for all is by doing them on the spot ;-) Let's hope we come to a consensus soon.

LKaemmerling replied:
As sample where it is with an singular alias too:
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vultr/vultr_server.py#L82

dagwieers replied:
We have many modules doing things we do not recommend or wouldn't recommend today. In any case there are more examples where we don't have singular parameter aliases for plural parameters (lists) than where we do it.

The most important reason for me is that it adds to the confusion, people expecting a single SSH key value is possible because the parameter used is singular. (Yes, they should have read the documentation, but lots of users don't and the singular form *is* deceiving)

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+                self.module.params.get("datacenter")
+            )
+        else:
+            self.module.fail_json(msg="Please specify a datacenter or location.")
```
If one of both is required, it is better to force this as part of the argspec.

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,408 @@
+    state: present
+
+# Remove a server
+- name: server should be absent
```
Please always capitalize the task names for consistency. (This ends up in the documentation as-si)
```suggestion
- name: Ensure the server is absent (remove if needed)
```

dagwieers reviewed (changes requested):
Various (mostly cosmetic) remarks.

LKaemmerling committed:
Apply review results from dagwieers

dagwieers commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,401 @@
+            ]
+
+        if (
+                self.module.params.get("location") is None
```
I was happy to see that pylint also disliked this formatting for conditional statements.
I know this is more a matter of taste, but we tend to do this for multiline conditional statements:
```python
    if self.module.params.get("location") is None and
            self.module.params.get("datacenter") is None:
        # When not given, the API will choose the location.
            params["location"] = None
            params["datacenter"] = None
```
or leave it on a single line (we accept lines up to 160 characters):
```python
    if self.module.params.get("location") is None and self.module.params.get("datacenter") is None:
```
And IMO this is more readable and readability is what we crave for ;-)

LKaemmerling committed:
Apply review results from dagwieers

LKaemmerling committed:
Remove ssh_key alias

gundalow commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,379 @@
+    id:
+        description:
+            - The ID of the Hetzner Cloud server to manage.
+            - Only required if no server I(name) is given
```
Does that mean `required_one_of=[['id', 'name']],` in the argspec?

LKaemmerling replied:
Yep, changed it.

gundalow commented on lib/ansible/plugins/doc_fragments/hcloud.py:
```diff
@@ -0,0 +1,27 @@
+        default: https://api.hetzner.cloud/v1
+        type: str
+requirements:
+  - python >= 2.6
```
Ansible 2.8 requires Python 2.6+, so this can be removed.
```suggestion
```

resmo commented on lib/ansible/modules/cloud/hcloud/hcloud_server.py:
```diff
@@ -0,0 +1,379 @@
+                },
+                **Hcloud.base_module_arguments()
+            ),
+            mutually_exclusive=[["location", "datacenter"]],
```
~~~diff
+           required_one_of=[['id', 'name']],
~~~

dagwieers reviewed (approved):
LGTM

LKaemmerling committed:
Apply review results from gundalow

resmo commented:
shipit

resmo commented:
Thanks!

gundalow commented:
Great to see this merged, thanks everybody for the reviews!

decentral1se commented:
Small docs fix over at https://github.com/ansible/ansible/pull/53466 :+1:
